lib/at_monitor/at_monitor.c

```diff
@@ -26,11 +26,14 @@ static K_FIFO_DEFINE(at_monitor_fifo);
 static K_HEAP_DEFINE(at_monitor_heap, CONFIG_AT_MONITOR_HEAP_SIZE);
 static K_WORK_DEFINE(at_monitor_work, at_monitor_task);
 
-static void at_monitor_dispatch(const char *notif)
+/* This is not static so that tests can call this function */
+void at_monitor_dispatch(const char *notif)
 {
 	struct at_notif_fifo *at_notif;
 	size_t sz_needed;
 
+	__ASSERT_NO_MSG(notif != NULL);
+
 	sz_needed = sizeof(struct at_notif_fifo) + strlen(notif) + sizeof(char);
 
 	at_notif = k_heap_alloc(&at_monitor_heap, sz_needed, K_NO_WAIT);
```

Comment:
If there is no space in the heap, we won't receive SMS to the SMS library, and we cannot ack it properly. This means SMS service will get stuck and you won't be able to receive or send messages. Normal SMS message is guaranteed to fit as `CONFIG_AT_MONITOR_HEAP_SIZE` is minimum 512 bytes.
However, if we receive a longer message, CMT notification is silently ignored and SMS service will get stuck. I found this issue with unit test `test_recv_fail_internal_pdu_buffer_too_short()` which is now commented out.
Then the question is do we need to be prepared for this. This is not something user should be able to trigger but it's the network that would send a malicious message being too long compared to what SMS can be based on 3GPP, and modem forwards it to app side. On the other hand, this is not something that's within the device's control so we should be functional no matter what network sends us.

If we wan to fix this, my only solution proposal would be to cut the AT notification to what fits into the buffer we have.
BTW, what was the reason for using own heap instead of static RAM for the notifications?

@lemrey What do you think?

Reply:
@trantanen Do you know what modem does if it receives a too long SMS? Just wondering whether that gets forwarded like that to the application side or not.

Reply:
I don't know how modem behaves in this case.

Reply:
When the workqueue task is executed, there could be any number of notifications pending to be dispatched. Reserving RAM statically for the notifications would not guarantee that we would have enough memory for all pending notifications at any time.

Trimming a notification is not too good, since it become unusable (depending on the notification?).
The thing is, any notification that is received exists somewhere in memory already-- the shared memory. So perhaps when we are unable to allocate on the heap to dispatch the notification, we could dispatch immediately in the IRQ and perhaps indicate that with a parameter in the monitor callback? Or maybe have a separate kind of monitor, one that gives the guarantee you'll never be out of memory for a notification, as long as you accept that it comes in a IRQ and have to handle it yourself.

Reply:
Ok, good that we have options. I don't really have strong opinions on which one to choose at this point.

Reply:
See #6402

Reply:
Content of https://github.com/nrfconnect/sdk-nrf/pull/6402 has been added to this PR to be able to take `AT_MONITOR_ISR` into use.

Reply:
PR https://github.com/nrfconnect/sdk-nrf/pull/6402 merged and rebased into this branch so this comment can be closed.